src/debug_flags.cc

```diff
@@ -12,10 +12,35 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <stdio.h>
+#include <map>
+#include <vector>
+#include <string>
+
+#include "graph.h"
+
 bool g_explaining = false;
 
 bool g_keep_depfile = false;
 
 bool g_keep_rsp = false;
 
 bool g_experimental_statcache = true;
+
+// Reasons each Node needs rebuilding, for "-d explain".
+typedef std::map<const Node*, std::vector<std::string> > Explanations;
+static Explanations explanations_;
```

Comment:
A few remarks: this is probably not the right place to put this, consider moving this to its own header + source files (e.g. explanations.h + explanations.cc), and declare a single Explanations class which Record() and Print() methods instead. Also an explanations_test.cc would be welcome to check that this works correctly even if the implementation changes in the future.

Also please avoid introducing new global variables. An Explanations* pointer should be passed to all functions that need to call the EXPLAIN() macro instead (I know this sucks, probably introduce this in its own commit). The Explanations object should probably be owned by the Builder class.

Reply:
And the pointer could be used for checks instead of the `g_explaining` global flag.

Reply:
Which global variable do you mean? If you're talking about `explanations_`, doesn't `static` make it local to the compilation unit? (It's quite possible I'm wrong -- it's been a long time since I've done C++).

Thanks for the review, by the way!

Reply:
Yes, I am talking about `explanations` which is global variable (in the sense of mutable global state). These are best avoided because they tend to hide dependencies from the source code, they have their own life-cycle which can introduce unexpected coupling issues, they make unit-testing more difficult, and they also complicate future refactors (e.g. there are plans to turn Ninja into a library, but it could be anything that tries to change the graph data structures); also they are generally not thread-safe. So let's not introduce more of them if possible :-)